src/Servers/Kestrel/Core/test/TlsListenerMiddlewareTests.cs

```diff
@@ -0,0 +1,563 @@
+        => RunTlsClientHelloCallbackTest_WithMultipleSegments(id, packets, nextMiddlewareShouldBeInvoked: true, tlsClientHelloCallbackExpected: true);
+
+    [Theory]
+    [MemberData(nameof(InvalidClientHelloData_Segmented))]
+    public Task OnTlsClientHelloAsync_InvalidData_MultipleSegments(int id, List<byte[]> packets)
+        => RunTlsClientHelloCallbackTest_WithMultipleSegments(id, packets, nextMiddlewareShouldBeInvoked: true, tlsClientHelloCallbackExpected: false);
+
+    [Fact]
+    public async Task RunTlsClientHelloCallbackTest_DeterministinglyReads()
+    {
+        var serviceContext = new TestServiceContext();
+
+        var pipe = new Pipe();
+        var writer = pipe.Writer;
+        var reader = new ObservablePipeReader(pipe.Reader);
+
+        var transport = new DuplexPipe(reader, writer);
+        var transportConnection = new DefaultConnectionContext("test", transport, transport);
+
+        var nextMiddlewareInvoked = false;
+        var tlsClientHelloCallbackInvoked = false;
+
+        var middleware = new TlsListenerMiddleware(
+            next: ctx =>
+            {
+                nextMiddlewareInvoked = true;
+                var readResult = ctx.Transport.Input.ReadAsync();
+                Assert.Equal(5, readResult.Result.Buffer.Length);
+
+                return Task.CompletedTask;
+            },
+            tlsClientHelloBytesCallback: (ctx, data) =>
+            {
+                tlsClientHelloCallbackInvoked = true;
+            }
+        );
+
+        await writer.WriteAsync(new byte[1] { 0x16 });
+        var middlewareTask = middleware.OnTlsClientHelloAsync(transportConnection);
+        await writer.WriteAsync(new byte[2] { 0x03, 0x01 });
+        await writer.WriteAsync(new byte[2] { 0x00, 0x20 });
+        await writer.CompleteAsync();
+
+        await middlewareTask;
+        Assert.True(nextMiddlewareInvoked);
+        Assert.False(tlsClientHelloCallbackInvoked);
+
+        // ensuring that we have read limited number of times
+        Assert.True(reader.ReadAsyncCounter is >= 2 && reader.ReadAsyncCounter is <= 3,
+            $"Expected ReadAsync() to happen about 2-3 times. Actually happened {reader.ReadAsyncCounter} times.");
+    }
+
+    private async Task RunTlsClientHelloCallbackTest_WithMultipleSegments(
+        int id,
+        List<byte[]> packets,
+        bool nextMiddlewareShouldBeInvoked,
+        bool tlsClientHelloCallbackExpected)
+    {
+        var pipe = new Pipe();
+        var writer = pipe.Writer;
+
+        var transport = new DuplexPipe(pipe.Reader, writer);
+        var transportConnection = new DefaultConnectionContext("test", transport, transport);
+
+        var nextMiddlewareInvokedActual = false;
+        var tlsClientHelloCallbackActual = false;
+
+        var fullLength = packets.Sum(p => p.Length);
+
+        var middleware = new TlsListenerMiddleware(
+            next: ctx =>
+            {
+                nextMiddlewareInvokedActual = true;
```

Comment:
We should probably check the length in the next middleware

Reply:
checked in case TLS client hello callback was invoked. In other case we dont know at which byte the invalid data failed, so we cant for sure check against the full packet length.